File: ports/cpp/src/core/sysiface.cpp
#include "sysiface.hpp"
#include "radiotap.hpp" // freq_to_channel

#include <algorithm>
#include <cctype>
#include <cstdlib>
#include <format>
#include <mutex>
#include <sstream>
#include <string>
#include <unordered_map>

#include <sys/wait.h>
#include <unistd.h>

namespace pf {

namespace {

struct CmdResult {
    bool ran = false;
    int exit_code = -1;
    std::string out;
};

// Run a command (argv style, no shell) and capture stdout.
CmdResult run_command(const std::vector<std::string>& args) {
    CmdResult res;
    if (args.empty())
        return res;

    int pipefd[2];
    if (pipe(pipefd) != 0)
        return res;

    pid_t pid = fork();
    if (pid < 0) {
        close(pipefd[0]);
        close(pipefd[1]);
        return res;
    }
    if (pid == 0) {
        dup2(pipefd[1], STDOUT_FILENO);
        dup2(pipefd[1], STDERR_FILENO);
        close(pipefd[0]);
        close(pipefd[1]);
        std::vector<char*> argv;
        argv.reserve(args.size() + 1);
        for (const auto& a : args)
            argv.push_back(const_cast<char*>(a.c_str()));
        argv.push_back(nullptr);
        execvp(argv[0], argv.data());
        _exit(127);
    }

    close(pipefd[1]);
    char buf[4096];
    ssize_t n;
    while ((n = read(pipefd[0], buf, sizeof(buf))) > 0)
        res.out.append(buf, static_cast<size_t>(n));
    close(pipefd[0]);

    int status = 0;
    waitpid(pid, &status, 0);
    res.ran = true;
    res.exit_code = WIFEXITED(status) ? WEXITSTATUS(status) : -1;
    return res;
}

std::string trim(std::string_view s) {
    size_t b = 0, e = s.size();
    while (b < e && std::isspace(static_cast<unsigned char>(s[b]))) ++b;
    while (e > b && std::isspace(static_cast<unsigned char>(s[e - 1]))) --e;
    return std::string(s.substr(b, e - b));
}

std::vector<std::string> split_ws(const std::string& s) {
    std::vector<std::string> out;
    std::istringstream iss(s);
    std::string tok;
    while (iss >> tok)
        out.push_back(tok);
    return out;
}

} // namespace

bool command_exists(std::string_view cmd) {
    static std::mutex mtx;
    static std::unordered_map<std::string, bool> cache;

    std::lock_guard lk(mtx);
    std::string key(cmd);
    if (auto it = cache.find(key); it != cache.end())
        return it->second;

    bool found = false;
    if (const char* path = std::getenv("PATH")) {
        std::string p(path);
        size_t start = 0;
        while (start <= p.size()) {
            size_t colon = p.find(':', start);
            std::string dir = p.substr(start, colon == std::string::npos ? colon : colon - start);
            if (!dir.empty()) {
                std::string full = dir + "/" + key;
                if (access(full.c_str(), X_OK) == 0) {
                    found = true;
                    break;
                }
            }
            if (colon == std::string::npos)
                break;
            start = colon + 1;
        }
    }
    cache.emplace(key, found);
    return found;
}

std::expected<void, std::string> set_channel_system(std::string_view iface, uint8_t channel) {
    std::string ch = std::to_string(channel);
    std::string ifs(iface);

#ifdef __APPLE__
    if (auto r = run_command(
            {"/System/Library/PrivateFrameworks/Apple80211.framework/Versions/"
             "Current/Resources/airport",
             ifs, "--channel", ch});
        r.ran) {
        if (r.exit_code != 0)
            return std::unexpected(std::format("airport returned status {}: {}", r.exit_code, trim(r.out)));
        return {};
    }
#endif

    if (command_exists("iw")) {
        auto r = run_command({"iw", "dev", ifs, "set", "channel", ch});
        if (!r.ran)
            return std::unexpected("failed to run iw");
        if (r.exit_code != 0)
            return std::unexpected(std::format("iw returned status {}: {}", r.exit_code, trim(r.out)));
        return {};
    }
    if (command_exists("iwconfig")) {
        auto r = run_command({"iwconfig", ifs, "channel", ch});
        if (!r.ran)
            return std::unexpected("failed to run iwconfig");
        if (r.exit_code != 0)
            return std::unexpected(std::format("iwconfig returned status {}: {}", r.exit_code, trim(r.out)));
        return {};
    }
    return std::unexpected("no supported tool found to set channel (need `iw` or `iwconfig`)");
}

std::optional<std::vector<std::pair<std::string, uint8_t>>>
scan_system_channels(std::string_view iface) {
    using Rows = std::vector<std::pair<std::string, uint8_t>>;
    std::string ifs(iface);

    if (command_exists("iw")) {
        auto r = run_command({"iw", "dev", ifs, "scan"});
        if (r.ran && r.exit_code == 0) {
            Rows rows;
            std::istringstream iss(r.out);
            std::string line, cur_bssid;
            bool have_bssid = false;
            while (std::getline(iss, line)) {
                std::string trimmed = line;
                if (size_t nb = trimmed.find_first_not_of(" \t"); nb != std::string::npos)
                    trimmed = trimmed.substr(nb);
                else
                    trimmed.clear();
                if (trimmed.starts_with("BSS ")) {
                    if (auto parts = split_ws(trimmed); parts.size() >= 2) {
                        cur_bssid = parts[1];
                        if (size_t paren = cur_bssid.find('('); paren != std::string::npos)
                            cur_bssid = cur_bssid.substr(0, paren);
                        have_bssid = true;
                    }
                }
                if (have_bssid && trimmed.starts_with("freq:")) {
                    if (auto parts = split_ws(trimmed); parts.size() >= 2) {
                        if (long freq = std::strtol(parts[1].c_str(), nullptr, 10); freq > 0)
                            if (auto c = freq_to_channel(static_cast<uint32_t>(freq)))
                                rows.emplace_back(cur_bssid, *c);
                    }
                    have_bssid = false;
                }
            }
            if (!rows.empty())
                return rows;
        }
    }

    if (command_exists("iwlist")) {
        auto r = run_command({"iwlist", ifs, "scanning"});
        if (r.ran && r.exit_code == 0) {
            Rows rows;
            std::istringstream iss(r.out);
            std::string line, cur_bssid;
            bool have_bssid = false;
            while (std::getline(iss, line)) {
                std::string t = trim(line);
                if (t.starts_with("Cell ") && t.find("Address:") != std::string::npos) {
                    size_t idx = t.find("Address:");
                    cur_bssid = trim(t.substr(idx + 8));
                    have_bssid = true;
                }
                if (have_bssid && t.starts_with("Channel:")) {
                    if (size_t colon = t.find(':'); colon != std::string::npos) {
                        std::string chs = trim(t.substr(colon + 1));
                        if (long c = std::strtol(chs.c_str(), nullptr, 10); c >= 0 && c <= 255)
                            rows.emplace_back(cur_bssid, static_cast<uint8_t>(c));
                    }
                    have_bssid = false;
                }
            }
            if (!rows.empty())
                return rows;
        }
    }
    return std::nullopt;
}

static std::expected<void, std::string> set_iface_type(std::string_view iface, std::string_view type) {
#if defined(__linux__)
    std::string ifs(iface), ty(type);
    if (!command_exists("ip") || !command_exists("iw"))
        return std::unexpected("`ip` or `iw` not found; cannot change interface mode");

    if (auto r = run_command({"ip", "link", "set", ifs, "down"}); !r.ran || r.exit_code != 0)
        return std::unexpected(std::format("ip down failed (status {})", r.exit_code));
    if (auto r = run_command({"iw", "dev", ifs, "set", "type", ty}); !r.ran || r.exit_code != 0)
        return std::unexpected(std::format("iw set type {} failed (status {})", ty, r.exit_code));
    if (auto r = run_command({"ip", "link", "set", ifs, "up"}); !r.ran || r.exit_code != 0)
        return std::unexpected(std::format("ip up failed (status {})", r.exit_code));
    return {};
#elif defined(__APPLE__)
    (void)iface; (void)type;
    return std::unexpected("automatic monitor-mode management is not implemented on macOS");
#else
    (void)iface; (void)type;
    return std::unexpected("monitor-mode management not supported on this OS");
#endif
}

std::expected<void, std::string> enable_monitor_mode(std::string_view iface) {
    return set_iface_type(iface, "monitor");
}
std::expected<void, std::string> disable_monitor_mode(std::string_view iface) {
    return set_iface_type(iface, "managed");
}

} // namespace pf

File: src/sysiface.rs
//! Debug logging + interaction with platform networking tools
//! (iw / iwconfig / iwlist / ip, and airport on macOS).
use std::collections::HashMap;
use std::process::Command;
use std::sync::{Mutex, OnceLock};

use crate::radiotap::freq_to_channel;

static DEBUG_FLAG: OnceLock<bool> = OnceLock::new();
static COMMAND_CHECK_CACHE: OnceLock<Mutex<HashMap<String, bool>>> = OnceLock::new();

pub fn debug_enabled() -> bool {
    *DEBUG_FLAG.get_or_init(|| {
        std::env::var("PF_DEBUG")
            .map(|v| {
                let v = v.trim();
                v == "1" || v.eq_ignore_ascii_case("true") || v.eq_ignore_ascii_case("yes")
            })
            .unwrap_or(false)
    })
}

pub fn log_debug(msg: &str) {
    if debug_enabled() {
        eprintln!("[DEBUG] {msg}");
    }
}

/// Check whether a program exists in PATH.
pub fn command_exists(cmd: &str) -> bool {
    let cache = COMMAND_CHECK_CACHE.get_or_init(|| Mutex::new(HashMap::new()));

    if let Ok(mut map) = cache.lock() {
        if let Some(&cached) = map.get(cmd) {
            return cached;
        }
        let found = which::which(cmd).is_ok();
        map.insert(cmd.to_string(), found);
        found
    } else {
        which::which(cmd).is_ok()
    }
}

/// Set wireless channel using platform-appropriate tools.
pub fn set_channel_system(iface: &str, channel: u8) -> Result<(), String> {
    if cfg!(target_os = "macos") {
        if let Ok(output) = Command::new("/System/Library/PrivateFrameworks/Apple80211.framework/Versions/Current/Resources/airport")
            .arg(iface)
            .arg("--channel")
            .arg(channel.to_string())
            .output()
        {
            if !output.status.success() {
                return Err(format!(
                    "airport returned status {} with stderr: {}",
                    output.status,
                    String::from_utf8_lossy(&output.stderr).trim()
                ));
            }
            return Ok(());
        }
    }

    if command_exists("iw") {
        let output = Command::new("iw")
            .arg("dev").arg(iface).arg("set").arg("channel").arg(channel.to_string())
            .output()
            .map_err(|e| format!("failed to run iw: {e}"))?;
        if !output.status.success() {
            return Err(format!(
                "iw returned status {} with stderr: {}",
                output.status,
                String::from_utf8_lossy(&output.stderr).trim()
            ));
        }
        return Ok(());
    }

    if command_exists("iwconfig") {
        let output = Command::new("iwconfig")
            .arg(iface).arg("channel").arg(channel.to_string())
            .output()
            .map_err(|e| format!("failed to run iwconfig: {e}"))?;
        if !output.status.success() {
            return Err(format!(
                "iwconfig returned status {} with stderr: {}",
                output.status,
                String::from_utf8_lossy(&output.stderr).trim()
            ));
        }
        return Ok(());
    }

    Err("no supported tool found to set channel (need `iw` or `iwconfig` on Linux, or `airport` on macOS)".into())
}

/// Scan for nearby APs and return (BSSID, channel) tuples using platform tools.
pub fn scan_system_channels(iface: &str) -> Option<Vec<(String, u8)>> {
    if cfg!(target_os = "macos") {
        if let Ok(output) = Command::new("/System/Library/PrivateFrameworks/Apple80211.framework/Versions/Current/Resources/airport")
            .arg("-s")
            .output()
        {
            if output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);
                let mut rows = Vec::new();
                for line in stdout.lines().skip(1) {
                    let tokens: Vec<&str> = line.split_whitespace().collect();
                    let mut bssid_idx = None;
                    for (i, t) in tokens.iter().enumerate() {
                        if t.len() == 17 && t.matches(':').count() == 5 {
                            bssid_idx = Some(i);
                            break;
                        }
                    }
                    let Some(idx) = bssid_idx else { continue };
                    if idx + 1 >= tokens.len() { continue; }
                    let bssid = tokens[idx].to_string();
                    if let Ok(ch) = tokens[idx + 1].parse::<u8>() {
                        rows.push((bssid, ch));
                    }
                }
                if !rows.is_empty() { return Some(rows); }
            }
        }
    }

    // `iw dev <iface> scan` (preferred on Linux)
    if command_exists("iw") {
        if let Ok(output) = Command::new("iw").arg("dev").arg(iface).arg("scan").output() {
            if output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);
                let mut rows = Vec::new();
                let mut cur_bssid: Option<String> = None;
                for line in stdout.lines() {
                    let trimmed = line.trim_start();
                    if trimmed.starts_with("BSS ") {
                        let parts: Vec<&str> = trimmed.split_whitespace().collect();
                        if parts.len() >= 2 {
                            let b = parts[1].split('(').next().unwrap_or(parts[1]);
                            cur_bssid = Some(b.trim().to_string());
                        }
                    }
                    if let Some(bssid) = &cur_bssid {
                        if trimmed.starts_with("freq:") {
                            if let Some(freq_s) = trimmed.split_whitespace().nth(1) {
                                if let Ok(freq) = freq_s.parse::<u32>() {
                                    if let Some(ch) = freq_to_channel(freq) {
                                        rows.push((bssid.clone(), ch));
                                    }
                                }
                            }
                            cur_bssid = None;
                        }
                    }
                }
                if !rows.is_empty() {
                    return Some(rows);
                }
            }
        }
    }

    // Fallback to `iwlist <iface> scanning`
    if command_exists("iwlist") {
        if let Ok(output) = Command::new("iwlist").arg(iface).arg("scanning").output() {
            if output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);
                let mut rows = Vec::new();
                let mut cur_bssid: Option<String> = None;
                for line in stdout.lines() {
                    let t = line.trim();
                    if t.starts_with("Cell ") && t.contains("Address:") {
                        if let Some(idx) = t.find("Address:") {
                            cur_bssid = Some(t[idx + 8..].trim().to_string());
                        }
                    }
                    if let Some(bssid) = &cur_bssid {
                        if t.starts_with("Channel:") {
                            if let Some(chs) = t.split(':').nth(1) {
                                if let Ok(ch) = chs.trim().parse::<u8>() {
                                    rows.push((bssid.clone(), ch));
                                }
                            }
                            cur_bssid = None;
                        }
                    }
                }
                if !rows.is_empty() {
                    return Some(rows);
                }
            }
        }
    }

    None
}

/// Enable monitor mode on the provided interface (Linux only).
pub fn enable_monitor_mode(iface: &str) -> Result<(), String> {
    set_iface_type(iface, "monitor")
}

/// Disable monitor mode (restore managed mode) on the provided interface (Linux only).
pub fn disable_monitor_mode(iface: &str) -> Result<(), String> {
    set_iface_type(iface, "managed")
}

fn set_iface_type(iface: &str, kind: &str) -> Result<(), String> {
    if cfg!(target_os = "linux") {
        if !command_exists("ip") || !command_exists("iw") {
            return Err("`ip` or `iw` not found; cannot change interface mode".into());
        }
        run_status(Command::new("ip").arg("link").arg("set").arg(iface).arg("down"), "ip down")?;
        run_status(Command::new("iw").arg("dev").arg(iface).arg("set").arg("type").arg(kind), "iw set type")?;
        run_status(Command::new("ip").arg("link").arg("set").arg(iface).arg("up"), "ip up")?;
        Ok(())
    } else if cfg!(target_os = "macos") {
        Err("automatic monitor-mode management is not implemented on macOS".into())
    } else {
        Err("monitor-mode management not supported on this OS".into())
    }
}

fn run_status(cmd: &mut Command, label: &str) -> Result<(), String> {
    let s = cmd.status().map_err(|e| format!("failed to run {label}: {e}"))?;
    if !s.success() {
        return Err(format!("{label} returned status {s}"));
    }
    Ok(())
}
